bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/ComponentLight.java

```diff
@@ -118,7 +118,7 @@ public ComponentLight(CFactory.ComponentConfiguration builder) {
                 .commandTopic(channelConfiguration.brightness_command_topic, channelConfiguration.retain)//
                 .build(false);
 
-        channels.put(switchChannelID, colorChannel);
+        channels.put(colorChannelID, colorChannel);
```

Comment:
Why is only the color channel added?

Reply:
Good question.  I'm not certain as this was designed by @davidgraeff .  Before my change, only the color channel was added, but with the light channel ID.  From my limited experimentation the color channel is the only one that seems to matter.  The OpenHab light Item in the UI has sliders for color, saturation, and brightness, but in the end the changes are translated into RGB color values and sent to the color channel.

Reply:
Mhm. I did look at the code and channels are always added with their id in `buildChannel`, in this case `channels.put(colorChannelID, colorChannel);` is already called in l. 114. 

Since none of the other components call `channels.put`, this probably is done by intention here. I'm not 100% sure it is a good idea to change that. Are there any problems that arise from the old code?

Reply:
That is interesting.  
What I found was that for a MQTT Homeassistant light, after it is discovered a Color item is created in Open HAB and a channel is created between the Color item and the Light Thing.  Then if I change any of the properties of the Color item in the OpenHAB Paper UI (color, brightness etc.) This method is called from line 125 of AbstractMQTTThingHandler in mqtt.generic:

    @Override
    public void handleCommand(ChannelUID channelUID, Command command) {
        if (connection == null) {
            return;
        }

        final @Nullable ChannelState data = getChannelState(channelUID);

        if (data == null) {
            logger.warn("Channel {} not supported!", channelUID);
            return;
        }

Debugging shows that the result from getChannelState was null, and the channel UID passed in is "color".

getChannelState is implemented here at line 223 of HomeAssistantThingHander.java

    @SuppressWarnings({ "null", "unused" })
    @Override
    public @Nullable ChannelState getChannelState(ChannelUID channelUID) {
        String groupID = channelUID.getGroupId();
        if (groupID == null) {
            return null;
        }
        AbstractComponent<?> component;
        synchronized (haComponents) { // sync whenever discoverComponents is started
            component = haComponents.get(groupID);
        }
        if (component == null) {
            return null;
        }
        CChannel componentChannel = component.channel(channelUID.getIdWithoutGroup());
        if (componentChannel == null) {
            return null;
        }

The component.channel function is implemented at line 185 in AbstractComponent.java
    public @Nullable CChannel channel(String channelID) {
        return channels.get(channelID);
    }

Further debugging shows that channels.get returns null without my changes, the debugger showed that the channels tree contained only one item labeled "light".  So a lookup with the name "color" failed.  With my change, registering the "color" UID in channels.put, the lookup succeeds and the changes made in the Color Item propagate through properly to the MQTT broker.  I can try some additional debugging to see why the put that you mention doesn't seem to be called.

Reply:
I haven't had a chance to debug further.  But looking into the implementation of buildChannel in AbstractComponent, this calls the constructor for CChannel.Builder, but it does not call the build() method on the returned "builder" item.  The constructor doesn't call the build() method either.  Thoughts?  Does the fix belong where I put it, or should buildChannel call the build() method on the newly created CChannel.Builder?

Reply:
Interesting. I didn't notice that you can pass `true` or `false` to the build. Ok. I think I understood why this is done. The reason is that homeassistant has many different topics where information about this device is posted. We need to create three channels to gather all information, but we do not want to expose them to openHAB, therfore we pass "addToComponet=false".

So the question is: why do we add the color channel with the id of the switch channel. I think this is wrong. @jochen314, since you did a lot of work in the whole MQTT binding in the last weeks, WDYT? I think this sounds correct.

Reply:
This fix looks correct to me.

I never liked this multiplexing done here and it is on my list to change. But HA is very complex here...